### PR TITLE
Disallow dimension reducing slicing of storages 

### PR DESCRIPTION
**Pull request by tehrengruber**
## Description

Disallow dimension reducing slicing of storage to prevent inconsistent metadata.

__Problem__
Taking a slice of a gt storage that results in a storage of lower dimension has incorrect metadata, e.g. origin, mask are just the "old" values. Fixing this properly is non trivial as it requires recognizing which dimension was collapsed. However the slices given to `__getitem__` are not accessible anymore in `__array_finalize__`, i.e. the point where the metadata is updated. Even in case we would write the machinery to do this, we would need to update all metadata such that it is equal to creating a new storage with the same data. Something that is error prone and time consuming given the current state of the storages.

Example from #31:
```
arr = gt4py.storage.empty("numpy", default_origin=[0, 0, 0], shape=[10, 10, 10], dtype=float)
s = arr[:, :, 0]
```

__Proposed solution__

In the light of advances to simplify storages in the past and the aforementioned problems I propose to disallow dimension reducing slicing of storages all together. For users that require this functionality using ~~`Storage.data`~~ `Storage.to_numpy()`, slicing and creating a new storage from the result provides a clean workaround. If this use case occurs frequently in user code we could provide a utility function or similar.

@jdahm What is your take on this? Does this happen frequently in FV3?

## Requirements

Before submitting this PR, please make sure:

- [x] You have run the code checks, tests and documentation build successfully
- [x] All fixes and all new functionality are tested and documentation is up to date
- [x] You looked at the [review checklist](https://github.com/GridTools/gt4py/blob/master/REVIEW_CHECKLIST.txt)

Additionally, if this PR contains code authored by new contributors:

- [x] All the authors are covered by a valid contributor assignment agreement,
signed by the employer if needed, provided to ETH Zurich
- [x] The names of all the new contributors have been added to an updated
version of the AUTHORS.rst file included in the PR

### CODE REVIEW

**Comment by jdahm** [comment]
Hi @tehrengruber, I like this idea. We discussed it this morning in the dev meeting and it looks as though fv3 uses a wrapper around the storage data anyway when it slices like that.

**Comment by tehrengruber** [comment]
bors try

File: docs/gt4py/quickstart.rst [hunk 1]
```diff
@@ -207,6 +207,10 @@ The ``default_origin`` parameter plays two roles:
 #. If when calling the stencil, no other `origin` is specified, this value is where the `iteration domain` begins, i.e.
    the grid point with the lowest index where a value is written.
 
+.. note::
+    Slicing a storage with reduced dimension, e.g. `field_a[:, :, 0]`, is not supported. Use `Storage.data` to retrieve
+    a slicable array, e.g. `field_a.data[:, :, 0]`, and create a new storage from it if needed.
```
Comment (jdahm) [inline docs/gt4py/quickstart.rst:212]: 👍

**Comment by tehrengruber** [comment]
bors try

**Review by jdahm — APPROVED** [review]
Looks right to me!

**Comment by tehrengruber** [comment]
bors try

File: src/gt4py/storage/storage.py [hunk 2]
```diff
@@ -207,6 +207,11 @@ def __array_finalize__(self, obj):
                     raise RuntimeError(
                         "Meta information can not be inferred when creating Storage views from other classes than Storage."
                     )
+                if self.ndim != obj.ndim and self.ndim != 0:
```
Comment (gronerl) [inline src/gt4py/storage/storage.py:210]: What case does the self.ndim != 0 cover? So it's still allowed to access scalar elements?

File: src/gt4py/storage/storage.py [hunk 3]
```diff
@@ -427,6 +437,18 @@ def _construct(cls, backend, dtype, default_origin, shape, alignment, layout_map
     def data(self):
         return self._device_field
 
+    def to_numpy(self, copy=False):
+        self.synchronize()
```
Comment (gronerl) [inline src/gt4py/storage/storage.py:441]:
```suggestion
        self.device_to_host()
```

File: src/gt4py/storage/storage.py [hunk 4]
```diff
@@ -427,6 +437,18 @@ def _construct(cls, backend, dtype, default_origin, shape, alignment, layout_map
...
+        if copy:
+            return copy.deepcopy(np.asarray(self))
+        return np.asarray(self)
+
+    def to_copy(self, copy=False):
+        self.synchronize()
```
Comment (gronerl) [inline src/gt4py/storage/storage.py:447]:
```suggestion
    def to_cupy(self, copy=False):
        self.host_to_device()
```
note also the typo in the method name

File: src/gt4py/storage/storage.py [hunk 5]
```diff
@@ -427,6 +437,18 @@ def _construct(cls, backend, dtype, default_origin, shape, alignment, layout_map
     def data(self):
         return self._device_field
 
+    def to_numpy(self, copy=False):
+        self.synchronize()
+        if copy:
+            return copy.deepcopy(np.asarray(self))
```
Comment (gronerl) [inline src/gt4py/storage/storage.py:443]:
```suggestion
            return np.array(self, subok=False)
```
Is the numpy way of doing copies. Maybe needs double checking if this doesn't result in a storage with our numpy patch-hack.

File: src/gt4py/storage/storage.py [hunk 6]
```diff
@@ -427,6 +437,18 @@ def _construct(cls, backend, dtype, default_origin, shape, alignment, layout_map
...
+        return np.asarray(self)
+
+    def to_copy(self, copy=False):
+        self.synchronize()
+        if copy:
+            return copy.deepcopy(cp.asarray(self))
```
Comment (gronerl) [inline src/gt4py/storage/storage.py:449]:
```suggestion
            return cp.array(self)
```

File: tests/test_unittest/test_storage.py [hunk 7]
```diff
@@ -944,7 +944,7 @@ def test_sum_gpu():
         shape=shape,
     )
 
-    q1[i1 : i2 + 1, jslice, 0] = cp.sum(q2[i1 : i2 + 1, jslice, :], axis=2)
+    q1[i1 : i2 + 1, jslice, 0] = cp.sum(q2.data[i1 : i2 + 1, jslice, :], axis=2)
```
Comment (gronerl) [inline tests/test_unittest/test_storage.py:947]: Looks like this test is supposed testing suming over storages (which is a ufunc) if we just use .data this test tests nothing - i.e. we should probably remove the test as it's just no longer supported to do sums.

**Review by gronerl — CHANGES_REQUESTED** [review]

File: src/gt4py/storage/storage.py [hunk 8]
```diff
@@ -427,6 +437,18 @@ def _construct(cls, backend, dtype, default_origin, shape, alignment, layout_map
...
+        return self.view(cp.asarray(self))
```
Comment (gronerl) [inline src/gt4py/storage/storage.py:450]:
```suggestion
        return cp.asarray(self)
```
self.view() will only work to upcast to the ndarray baseclass or possibly other np.ndarray subclasses, which cp.ndarray is not.

File: src/gt4py/storage/storage.py (hunk 2 above)
Comment (tehrengruber) [reply src/gt4py/storage/storage.py:210]: This is not 100% clean, but I would say the benefits of allowing scalar storages (with incorrect metadata) overweight here. For example `numpy.allclose` relies on reducing an array/storage of booleans to a scalar boolean, so disallowing this case would mean require all occurences.

**Comment by gronerl** [comment]
bors try

**Review by gronerl — APPROVED** [review]

**Pull request merged**